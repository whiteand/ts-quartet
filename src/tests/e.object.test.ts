import { e as v } from "..";
import { testValidatorImpure } from "./testValidatorImpure";

describe("v({ ... })", () => {
  test("v({})", () => {
    const notNull = v({});
    testValidatorImpure(
      notNull,
      [{}, 1, 0, false, true, "", []],
      [null, undefined]
    );
  });
  test("v({ a: number })", () => {
    const notNull = v({ a: v.number });
    testValidatorImpure(
      notNull,
      [{ a: NaN }, { a: Infinity }, { a: 1, b: "123" }],
      [null, undefined, {}, 1, 0, false, true, "", []]
    );
  });
  test('v({ a: number, [restOmit]: ["valid"] })', () => {
    const notNull = v({ a: v.number, [v.restOmit]: ["valid"] });
    testValidatorImpure(
      notNull,
      [{ a: NaN }, { a: Infinity }, { a: 1 }],
      [null, undefined, {}, 1, 0, false, true, "", []]
    );
  });
  test("v({ [restOmit]: something })", () => {
    const notNull = v({
      [v.restOmit]: ["andrew"]
    });
    testValidatorImpure(
      notNull,
      [{}, 1, 0, false, true, "", []],
      [null, undefined]
    );
  });
  test("v({ [rest]: number })", () => {
    const notNull = v({
      [v.rest]: v.number
    });
    testValidatorImpure(
      notNull,
      [{}, 1, 0, false, true, "", [], { a: 1 }],
      [{ b: "string" }, null, undefined]
    );
  });

  test('v({ [rest]: number, [restOmit]: ["valid"] })', () => {
    const notNull = v({
      [v.rest]: v.number,
      [v.restOmit]: ["valid"]
    });
    testValidatorImpure(
      notNull,
      [
        {},
        1,
        0,
        false,
        true,
        "",
        [],
        { a: 1 },
        { a: 1, valid: null },
        { valid: null }
      ],
      [{ b: "string" }, null, undefined]
    );
  });
  test("v({ a: string, [rest]: number })", () => {
    const notNull = v({
      a: v.string,
      [v.rest]: v.number
    });
    testValidatorImpure(
      notNull,
      [{ a: "" }, { a: "1" }, { a: "1", b: 1 }],
      [
        { b: "string" },
        null,
        undefined,
        { a: 1 },
        {},
        [],
        { a: null },
        { a: "1", b: "1" }
      ]
    );
  });
  test('v({ [rest]: number, a: string, [restOmit]: ["b"] })', () => {
    const notNull = v({
      a: v.string,
      [v.rest]: v.number,
      [v.restOmit]: ["b"]
    });
    testValidatorImpure(
      notNull,
      [
        { a: "" },
        { a: "1" },
        { a: "1", b: 1 },
        { a: "1", b: 1, c: 2 },
        { a: "1", b: null },
        { a: "1", b: 2 }
      ],
      [{ b: "string" }, null, undefined, { a: 1 }, {}, [], { a: null }]
    );
  });
  test("{ v.pair }", () => {
    let pairFromValidator: any | null = null;
    const checkDict = v({
      [v.rest]: v.pair(
        v.custom(pair => {
          pairFromValidator = pair;
          return true;
        })
      )
    });
    expect(
      checkDict({
        a: 123
      })
    ).toBe(true);
    expect(pairFromValidator).toEqual({ key: "a", value: 123 });
  });
  test("{ v.pair }", () => {
    let pairFromValidator: any | null = null;
    const checkDict = v({
      a: v.pair(
        v.custom(pair => {
          pairFromValidator = pair;
          return true;
        })
      )
    });
    expect(
      checkDict({
        a: 123
      })
    ).toBe(true);
    expect(pairFromValidator).toEqual({ key: "a", value: 123 });
  });
  test("symbol prop", () => {
    const checkQuartet = v({
      [Symbol.for("quartet")]: v.number
    });
    testValidatorImpure(
      checkQuartet,
      [
        { [Symbol.for("quartet")]: 1 },
        { [Symbol.for("quartet")]: "1" },
        {},
        0,
        1,
        false,
        true
      ],
      [null, undefined]
    );
  });
});
