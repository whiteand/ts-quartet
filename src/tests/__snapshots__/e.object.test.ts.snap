// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v({ ... }) symbol prop 1`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "propsSchemas": Object {},
      "type": "Object",
    },
    "value": null,
  },
]
`;

exports[`v({ ... }) symbol prop 2`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "propsSchemas": Object {},
      "type": "Object",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ [rest]: number }) 1`] = `
Array [
  Object {
    "path": Array [
      "b",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": "string",
  },
]
`;

exports[`v({ ... }) v({ [rest]: number }) 2`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "[v.restOmit]": Array [],
      "[v.rest]": Object {
        "type": "Number",
      },
      "propsSchemas": Object {},
      "type": "Object",
    },
    "value": null,
  },
]
`;

exports[`v({ ... }) v({ [rest]: number }) 3`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "[v.restOmit]": Array [],
      "[v.rest]": Object {
        "type": "Number",
      },
      "propsSchemas": Object {},
      "type": "Object",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ [rest]: number, [restOmit]: ["valid"] }) 1`] = `
Array [
  Object {
    "path": Array [
      "b",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": "string",
  },
]
`;

exports[`v({ ... }) v({ [rest]: number, [restOmit]: ["valid"] }) 2`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "[v.restOmit]": Array [
        "valid",
      ],
      "[v.rest]": Object {
        "type": "Number",
      },
      "propsSchemas": Object {},
      "type": "Object",
    },
    "value": null,
  },
]
`;

exports[`v({ ... }) v({ [rest]: number, [restOmit]: ["valid"] }) 3`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "[v.restOmit]": Array [
        "valid",
      ],
      "[v.rest]": Object {
        "type": "Number",
      },
      "propsSchemas": Object {},
      "type": "Object",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ [rest]: number, a: string, [restOmit]: ["b"] }) 1`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "String",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ [rest]: number, a: string, [restOmit]: ["b"] }) 2`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "[v.restOmit]": Array [
        "b",
      ],
      "[v.rest]": Object {
        "type": "Number",
      },
      "propsSchemas": Object {
        "a": Object {
          "type": "String",
        },
      },
      "type": "Object",
    },
    "value": null,
  },
]
`;

exports[`v({ ... }) v({ [rest]: number, a: string, [restOmit]: ["b"] }) 3`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "[v.restOmit]": Array [
        "b",
      ],
      "[v.rest]": Object {
        "type": "Number",
      },
      "propsSchemas": Object {
        "a": Object {
          "type": "String",
        },
      },
      "type": "Object",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ [rest]: number, a: string, [restOmit]: ["b"] }) 4`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "String",
    },
    "value": 1,
  },
]
`;

exports[`v({ ... }) v({ [rest]: number, a: string, [restOmit]: ["b"] }) 5`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "String",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ [rest]: number, a: string, [restOmit]: ["b"] }) 6`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "String",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ [rest]: number, a: string, [restOmit]: ["b"] }) 7`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "String",
    },
    "value": null,
  },
]
`;

exports[`v({ ... }) v({ [restOmit]: something }) 1`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "propsSchemas": Object {},
      "type": "Object",
    },
    "value": null,
  },
]
`;

exports[`v({ ... }) v({ [restOmit]: something }) 2`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "propsSchemas": Object {},
      "type": "Object",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number }) 1`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "propsSchemas": Object {
        "a": Object {
          "type": "Number",
        },
      },
      "type": "Object",
    },
    "value": null,
  },
]
`;

exports[`v({ ... }) v({ a: number }) 2`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "propsSchemas": Object {
        "a": Object {
          "type": "Number",
        },
      },
      "type": "Object",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number }) 3`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number }) 4`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number }) 5`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number }) 6`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number }) 7`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number }) 8`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number }) 9`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number, [restOmit]: ["valid"] }) 1`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "propsSchemas": Object {
        "a": Object {
          "type": "Number",
        },
      },
      "type": "Object",
    },
    "value": null,
  },
]
`;

exports[`v({ ... }) v({ a: number, [restOmit]: ["valid"] }) 2`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "propsSchemas": Object {
        "a": Object {
          "type": "Number",
        },
      },
      "type": "Object",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number, [restOmit]: ["valid"] }) 3`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number, [restOmit]: ["valid"] }) 4`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number, [restOmit]: ["valid"] }) 5`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number, [restOmit]: ["valid"] }) 6`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number, [restOmit]: ["valid"] }) 7`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number, [restOmit]: ["valid"] }) 8`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: number, [restOmit]: ["valid"] }) 9`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: string, [rest]: number }) 1`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "String",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: string, [rest]: number }) 2`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "[v.restOmit]": Array [],
      "[v.rest]": Object {
        "type": "Number",
      },
      "propsSchemas": Object {
        "a": Object {
          "type": "String",
        },
      },
      "type": "Object",
    },
    "value": null,
  },
]
`;

exports[`v({ ... }) v({ a: string, [rest]: number }) 3`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "[v.restOmit]": Array [],
      "[v.rest]": Object {
        "type": "Number",
      },
      "propsSchemas": Object {
        "a": Object {
          "type": "String",
        },
      },
      "type": "Object",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: string, [rest]: number }) 4`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "String",
    },
    "value": 1,
  },
]
`;

exports[`v({ ... }) v({ a: string, [rest]: number }) 5`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "String",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: string, [rest]: number }) 6`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "String",
    },
    "value": undefined,
  },
]
`;

exports[`v({ ... }) v({ a: string, [rest]: number }) 7`] = `
Array [
  Object {
    "path": Array [
      "a",
    ],
    "schema": Object {
      "type": "String",
    },
    "value": null,
  },
]
`;

exports[`v({ ... }) v({ a: string, [rest]: number }) 8`] = `
Array [
  Object {
    "path": Array [
      "b",
    ],
    "schema": Object {
      "type": "Number",
    },
    "value": "1",
  },
]
`;

exports[`v({ ... }) v({}) 1`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "propsSchemas": Object {},
      "type": "Object",
    },
    "value": null,
  },
]
`;

exports[`v({ ... }) v({}) 2`] = `
Array [
  Object {
    "path": Array [],
    "schema": Object {
      "propsSchemas": Object {},
      "type": "Object",
    },
    "value": undefined,
  },
]
`;
